### ISSUE TITLE
Private templates

### ISSUE THREAD

**Issue by akhobov** [go-swagger/go-swagger#1606]
## Problem statement

There is a possibility to provide custom templates to the go-swagger tool to customise generated code. This functionality can be useful to patch/workaround some issues in the generated code too. However not all templates can be overloaded as majority of them marked as private.

Do you think it would be good to lift that limitation so all templates can be customised? This will give API developers a possibility to workaround some issues while waiting for a new release that fix the original issues. Obviously they would evaluate the risks and remove that customisation during the upgrade to the next version.

**Comment by fredbi** [comment]
you can lift the constraint yourself by removing the "locked" attribute in `template_repo.go`

**Comment by akhobov** [comment]
Yes, I can. However that would require a custom build that is not a very handy options for a group of developers. That's the reason we are sticking to released versions rather than master - to ensure a homogeneous development environment.

I'm not insisting on that change. Just asking if that seems a good idea from your point of view.

**Comment by fredbi** [comment]
Some templates are locked because they are deeply intertwined with the rest of the generation code.
So customization here equates to a major fork.
A better strategy in the case of bugs in this case is to stick on master rather than releases and participate the contribution effort rather than waiting for a release.

**Comment by fredbi** [comment]
CI testing on master have been largely extended and there is no real quality benefit to expect from releases besides repo tagging.

**Comment by akhobov** [comment]
That's not the quality I'm talking about. Imaging there is a new dev joins your team that needs to install the tool or one of the developers decided to update the tooling while the rest continue to use their old versions. The generated code would be different depending on who run the generator. With tags there is some level of certainty that all devs are using the same version and as such they all can get the same result out of the go-swagger generator.

**Comment by fredbi** [comment]
I am just like you guys.
I have a specific branch on my fork to stick my other devs.
This branch is regularly sync'ed with master, or sometimes completely reworked when I need something special ahead of master (or against @casualjim's wishes... 🤣 ).

I m using `dep` and `source` in `Gopkg.toml` to control how dependent packages depend on all this.

Another strategy I put in place for some features (for instance, I needed a custom swagger serve command) is to bake my own docker image from a forked branch and distribute this independently.

**Comment by akhobov** [comment]
Got you. Thanks a lot!

Anyway, if you think that idea is not good enough you can close this issue and we would workaround it in a way similar to what you described even though it is not our favourite solution :)

**Comment by fredbi** [comment]
I think the discussion thread interesting and it is okay to keep it open and collect the opinions from other people.

But I am not the one who will change this. I'd rather people to discuss and contribute with their issues on the core generator (templates and model.go/operation.go) than twisting their custom things on locked templates. I can tell you I have a lot of custom templates for my personal devs,
but I would not build my own, say, schemavalidator.gotmpl by forking away. That's actually how I came to contribute so much to this repo.

What is really needed in my opinion to get more stable builds is to tag the go-openapi repos.

**Comment by casualjim** [comment]
are the fixes generally useful? why not just contribute them back?

**Comment by akhobov** [comment]
Due to the period of time it took 0.14.0 to be released I backported a couple of fixes to 0.13.0. One of them was in the support for the CSV format in query params. Sorry, can't remember details of the other one from the top of my head. Also considering a workaround for #1601 while a proper fix is being worked on.

**Comment by fredbi** [comment]
@akhobov with release 0.16 we have introduced the possibility to get onboard "contributed templates". A team contributed one such set of templates. Might provide some new ways to bring customization to the table and share.
